api/database/database.go

```diff
@@ -39,11 +41,34 @@ func SetupDatabase() (*gorm.DB, error) {
 	config.Logger = logger.Default.LogMode(logger.Info)
 
 	db, err := gorm.Open(mysql.Open(address.String()), &config)
-	if err != nil {
-		return nil, errors.Wrap(err, "could not connect to database")
+	sqlDB, dbErr := db.DB()
+	if dbErr != nil {
+		log.Println(dbErr)
+		return nil, dbErr
 	}
 
-	// TODO: Add connection retries
+	sqlDB.SetMaxOpenConns(80)
+
+	if err != nil {
+		ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
+		defer cancel()
+
+		for retryCount := 1; retryCount <= 5; retryCount++ {
+			select {
+			case <-ctx.Done():
+				log.Println(ctx.Err())
+				return nil, err
+			default:
+				if err := sqlDB.PingContext(ctx); err != nil {
+					log.Printf("WARN: Could not ping database: %s, Will retry after 1 second", err)
+					time.Sleep(time.Second)
+				} else {
+					return db, nil
+				}
+			}
+		}
+		return nil, err
+	}
```

Comment:
I think that we should use a new context for each attempt, instead of reusing the same for all attempts. I.e. move the creation of ctx inside of the loop, and replace the whole switch statement with the contents of just the default branch, (the .PingContext function will automatically stop and return an error if the timeout is reached).

Also a nice improvement would be to double the wait time for each attempt. For example the first time docker-compose is started it takes some time to setup the database (more than 5 seconds) this would allow it to wait for longer without spamming the pings each second.

I've added a suggestion to implement it below, what do you think?

```suggestion
	if err != nil {
		for retryCount := 1; retryCount <= 5; retryCount++ {
			ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
			defer cancel()

			waitSeconds := int(math.Pow(2, float64(retryCount-1)))

			if err := sqlDB.PingContext(ctx); err == nil {
				return db, nil
			}

			log.Printf("WARN: Could not ping database: %s. Will retry after %d second\n", err, waitSeconds)
			time.Sleep(time.Duration(waitSeconds) * time.Second)
		}
		return nil, err
	}
```